pkg/cluster/fixetcd.go

```diff
@@ -0,0 +1,517 @@
+	degradedEndpoints := m.checkEtcds(ctx, etcd)
+	if degradedEndpoints {
+		pods, err := m.kubernetescli.CoreV1().Pods(nameSpace).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+		de := m.comparePodEnvToIP(pods)
+		// Exit early if degradedEtcd is empty
+		// no further actions should be taken
+		if de.Node == "" {
+			return errors.New("degradedEtcd is empty, unable to remediate etcd deployment")
+		}
+		m.log.Infof("Found degraded endpoint: %v", de)
+
+		err = m.backupEtcdData(ctx, de.Node)
+		if err != nil {
+			return err
+		}
+
+		err = m.fixPeers(ctx, de, pods)
+		if err != nil {
+			return err
+		}
+
+		err = m.patchEtcd(ctx, etcd, []byte(patchEnableOveride))
+		if err != nil {
+			return err
+		}
+
+		err = m.deleteSecrets(ctx, de)
+		if err != nil {
+			return err
+		}
+
+		err = m.patchEtcd(ctx, etcd, []byte(fmt.Sprintf(`{"spec": {"forceRedeploymentReason": "single-master-recovery-%s"}}`, time.Now())))
+		if err != nil {
+			return err
+		}
+
+		err = m.patchEtcd(ctx, etcd, []byte(patchDisableOverides))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *manager) patchEtcd(ctx context.Context, e *operatorv1.Etcd, patch []byte) error {
+	_, err := m.operatorcli.OperatorV1().Etcds().Patch(ctx, e.Name, types.MergePatchType, patch, metav1.PatchOptions{})
+	if err != nil {
+		return err
+	}
+	m.log.Infof("Patched etcd %s", e.Name)
+
+	return nil
+}
+
+func (m *manager) deleteSecrets(ctx context.Context, de *degradedEtcd) error {
+	secrets, err := m.kubernetescli.CoreV1().Secrets(nameSpace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, s := range secrets.Items {
+		b, err := regexp.MatchString("."+de.Node+"$", s.Name)
+		if err != nil {
+			return nil
+		}
+
+		if b {
+			err = m.kubernetescli.CoreV1().Secrets(nameSpace).Delete(ctx, s.Name, metav1.DeleteOptions{})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (m *manager) getPeerPods(pods []corev1.Pod, de *degradedEtcd) (string, error) {
+	regNode, err := regexp.Compile(`.master-[0-9]$`)
+	if err != nil {
+		return "", err
+	}
+	// regPod, err := regexp.Compile(`^etcd-[a-z]*-cluster-[0-9A-Za-z]*-master-[0-9]$`)
+	regPod, err := regexp.Compile(`etcd-` + m.doc.OpenShiftCluster.Name + `-[0-9A-Za-z]*-master-[0-9]$`)
+	if err != nil {
+		return "", err
+	}
+
+	var peerPods string
+	for _, p := range pods {
+		if regNode.MatchString(p.Spec.NodeName) && regPod.MatchString(p.Name) && p.Name != de.Pod {
+			peerPods += p.Name + " "
+		}
+	}
+	return peerPods, nil
+}
+
+func (m *manager) fixPeers(ctx context.Context, de *degradedEtcd, pods *corev1.PodList) error {
+	peerPods, err := m.getPeerPods(pods.Items, de)
+	if err != nil {
+		return err
+	}
+
+	_, _, _, scc, err := m.createPrivilegedServiceAccount(ctx, serviceAccountName, kubeServiceAccount)
+	if err != nil {
+		return err
+	}
+
+	name := jobName + "-fix-peers"
+	b, err := m.kubernetescli.BatchV1().Jobs(nameSpace).Create(ctx, &batchv1.Job{
```

Comment:
There's one bigger concern I have, which is that using a PUCM to start jobs that consume a script in-cluster is not really a pattern we have in PUCM right now, and is error prone. Usually the PUCM/adminUpdate flow is to have RP make the changes directly, and then wait. The pattern in-use here is to use RP to create jobs in-cluster, and then wait on those jobs. We're relying on pods and service accounts being created and deleted, images being pulled, and a bash script, which then goes and changes things, and it might be too complicated or introduces too many failure points.

If it's doable I think we should have RP do the job of the script, instead of committing the bash and making a job to execute it. I *think* that this will be possible based on my understanding of other PUCM steps, and the way that they modify k8s resources and then wait (the PUCM steps that rotate certs or change openshift crds are a good example of this)

Reply:
I'll have to review some examples and see if it's possible. If the same thing is done in other PUCM steps then I would think so.